### DISCUSSION TITLE
vertico-directory-enter, vertico-preselect 'first, and directory completion

### DISCUSSION THREAD

**Discussion by Niluge-KiWi** [minad/vertico#411] [General]
(new user of vertico here, having declared .emacs bankruptcy after 15 years of cruft accumulation and stagnation)

Migrating for ido, my muscle memory for `C-x C-f RET` expected continue completion in first sub-directory candidate.
Instead, by default, vertico opens dired on current directory, unexpectedly for my muscle memory.

So I started using `vertico-directory-enter` mapped on `RET` and switched `vertico-preselect` to `first`.

But then I hit a new issue: when selecting a directory (e.g. via `rg`), I would expect `RET` to select the current directory (like other inputs asked by `rg`), but instead I get:
- 2 candidates: `./`, `../`
- preselect on first candidate: `./`
- `RET` on `./` does simply nothing; in fact `vertico-directory-enter` enters into `./` directory, and cleans the path by removing useless `/./`, which is effectively a no-op, with zero visual indication that something happened

I then read more:
- maybe losely related to https://github.com/minad/vertico/tree/main#submitting-the-empty-string?
- https://github.com/minad/vertico/issues/302 and https://github.com/minad/vertico/issues/306
- dug more in vertico code

An easy reproduction is: `(completing-read-default "In directory: " 'read-file-name-internal 'file-directory-p t "~/.emacs.d/elpa/vertico-20230930.1918/" 'file-name-history "~/.emacs.d/elpa/vertico-20230930.1918/" nil)`

Here is my current analysis of the situation:

=> for `vertico-directory-enter` users, `vertico-preselect` `first` seems best, but we would need to:
1. either get rid of `./`: when first candidate is `./`, i.e. the same as prompt, don't (semantically) duplicate things: only in prompt, and thus preselect the prompt (though contrary to the current semantic of `vertico-preselect` `first`...)
2. or make `vertico-directory-enter` *not* enter (continue completion into) directory if selection is `./`, i.e. if it would result in noop; and thus do the alternate thing: `vertico-exit`
3. or make `vertico-preselect` `directory-(candidate-as-prompt-)unless-completing-directories`? (and keep the semantic duplicate entries between prompt & candidate)
4. other idea?

-> 1. is hard for me to implement; because I could not understand how './' is generated (but I'd like to know more :) )
-> 2. is easy for me to implement quick and dirty: add `((not (string-suffix-p "/./" cand)))` to `vertico-directory-enter` `if-let` check, just after defining `cand`, it works!
-> 3. seems overly complicated for a user option semantic

if we want a non-quick-and-dirty solution it seems to be best to me to look into the duplicate selections issues of 1.:
It's the first time I encounter a (semantic) duplicate: e.g. for `find-file` we never seem to have a candidate redundant with the prompt. Maybe this merit more thinking there?

What do you think?
If 2. is OK I can open a PR.

Thanks

**Comment by minad** [comment]
> It's the first time I encounter a (semantic) duplicate: e.g. for find-file we never seem to have a candidate redundant with the prompt. 

Where do you observe this issue? Only in rg.el? Maybe it could be fixed there (solution 1). Otherwise 2 sounds like a reasonable solution.

**Reply by Niluge-KiWi** [reply]
> Where do you observe this issue? Only in rg.el?

I've for now only seen this in rg.el for directory selection.

The stacktrace:
```
Debugger entered--entering a function:
* vertico-directory-enter(nil)
  funcall-interactively(vertico-directory-enter nil)
  command-execute(vertico-directory-enter)
  #<subr completing-read-default>("In directory: " read-file-name-internal file-directory-p t "~/.emacs.d/elpa/vertico-20230930.1918/" file-name-history "~/.emacs.d/elpa/vertico-20230930.1918/" nil)
  apply((#<subr completing-read-default> "In directory: " read-file-name-internal file-directory-p t "~/.emacs.d/elpa/vertico-20230930.1918/" file-name-history "~/.emacs.d/elpa/vertico-20230930.1918/" nil))
  vertico--advice(#<subr completing-read-default> "In directory: " read-file-name-internal file-directory-p t "~/.emacs.d/elpa/vertico-20230930.1918/" file-name-history "~/.emacs.d/elpa/vertico-20230930.1918/" nil)
  apply(vertico--advice #<subr completing-read-default> ("In directory: " read-file-name-internal file-directory-p t "~/.emacs.d/elpa/vertico-20230930.1918/" file-name-history "~/.emacs.d/elpa/vertico-20230930.1918/" nil))
  completing-read-default("In directory: " read-file-name-internal file-directory-p t "~/.emacs.d/elpa/vertico-20230930.1918/" file-name-history "~/.emacs.d/elpa/vertico-20230930.1918/" nil)
  read-file-name-default("In directory: " "/home/niluge/.emacs.d/elpa/vertico-20230930.1918/" "/home/niluge/.emacs.d/elpa/vertico-20230930.1918/" t nil file-directory-p)
  read-file-name("In directory: " "/home/niluge/.emacs.d/elpa/vertico-20230930.1918/" "/home/niluge/.emacs.d/elpa/vertico-20230930.1918/" t nil file-directory-p)
  read-directory-name("In directory: " nil "/home/niluge/.emacs.d/elpa/vertico-20230930.1918/" t)
  byte-code("\301\302!\303 \304\305\306\10\302$E\207" [default-directory rg-read-pattern t rg-read-files read-directory-name "In directory: " nil] 7)
  #<subr rg-literal--transient>()
  apply(#<subr rg-literal--transient> nil)
  #f(compiled-function (fn &rest args) #<bytecode 0x172507923aa5eca>)(#<subr rg-literal--transient>)
  apply(#f(compiled-function (fn &rest args) #<bytecode 0x172507923aa5eca>) #<subr rg-literal--transient> nil)
  (lambda (fn &rest args) (interactive #f(compiled-function (spec) #<bytecode 0xd19e79bd7895f5b>)) (apply '#f(compiled-function (fn &rest args) #<bytecode 0x172507923aa5eca>) fn args))(#<subr rg-literal--transient>)
  apply((lambda (fn &rest args) (interactive #f(compiled-function (spec) #<bytecode 0xd19e79bd7895f5b>)) (apply '#f(compiled-function (fn &rest args) #<bytecode 0x172507923aa5eca>) fn args)) #<subr rg-literal--transient> nil)
  rg-literal--transient()
  funcall-interactively(rg-literal--transient)
  command-execute(rg-literal--transient)
```

The doc for `completing-read-default` says it's deprecated to call with INITIAL-INPUT and recommend using DEF instead. We use both.

Upper frame: rg actuall calls `(read-directory-name "In directory: " nil "/home/niluge/.emacs.d/elpa/vertico-20230930.1918/" t)`
Calling `(read-directory-name "In directory: " nil nil t)` doesn't change anything.

I'm not sure what's wrong in rg.el; if we should change things there why not, but maybe vertico (and vertico-directory) should anyway handle such case: read-directory-name ending (or even starting) in a directory with no subdir, and selecting it.

**Comment by minad** [comment]
On 11/19/23 19:40, Thomas R. wrote:
> I'm not sure what's wrong in rg.el; if we should change things there why
> not, but maybe vertico (and vertico-directory) should anyway handle such
> case: read-directory-name ending (or even starting) in a directory with
> no subdir, and selecting it.

I pushed a fix. Please check if it works better now.

**Reply by Niluge-KiWi** [reply]
I confirm this works with 0115a6e01d143e5b6ce00a6ce8eb9c8a309e4067, thank you!
